store/auth.go

```diff
@@ -26,218 +26,157 @@ import (
-		return decodeJSONBody(resp, success, failure)
-	})
-}
+// deviceAuthorizer authorizes requests using devicew or user
+// managed via the DeviceAndAuthContext.
+// This is the default authorizer used by Store if a
+// DeviceAndAuthContext is supplied.
+type deviceAuthorizer struct {
+	UserAuthorizer
 
-// retryPostRequest calls doRequest and decodes the response in a retry loop.
-func retryPostRequest(httpClient *http.Client, endpoint string, headers map[string]string, data []byte, readResponseBody func(resp *http.Response) error) (*http.Response, error) {
-	return httputil.RetryRequest(endpoint, func() (*http.Response, error) {
-		req, err := http.NewRequest("POST", endpoint, bytes.NewBuffer(data))
-		if err != nil {
-			return nil, err
-		}
-		for k, v := range headers {
-			req.Header.Set(k, v)
-		}
+	endpointURL func(p string, query url.Values) *url.URL
 
-		return httpClient.Do(req)
-	}, readResponseBody, defaultRetryStrategy)
+	sessionMu sync.Mutex
 }
 
-// requestStoreMacaroon requests a macaroon for accessing package data from the ubuntu store.
-func requestStoreMacaroon(httpClient *http.Client) (string, error) {
-	const errorPrefix = "cannot get snap access permission from store: "
-
-	data := map[string]interface{}{
-		"permissions": []string{"package_access", "package_purchase"},
+func (a *deviceAuthorizer) Authorize(r *http.Request, dauthCtx DeviceAndAuthContext, user *auth.UserState, opts *AuthorizeOptions) error {
+	var firstError error
+	if opts.deviceAuth {
+		if device, err := dauthCtx.Device(); err == nil && device != nil && device.SessionMacaroon != "" {
+			r.Header.Set(hdrSnapDeviceAuthorization[opts.apiLevel], fmt.Sprintf(`Macaroon root="%s"`, device.SessionMacaroon))
+		} else if err != nil {
+			firstError = err
+		}
 	}
 
-	var err error
-	macaroonJSONData, err := json.Marshal(data)
-	if err != nil {
-		return "", fmt.Errorf(errorPrefix+"%v", err)
+	if err := a.UserAuthorizer.Authorize(r, dauthCtx, user, opts); err != nil && firstError == nil {
+		firstError = err
 	}
+	return firstError
+}
 
-	var responseData struct {
-		Macaroon string `json:"macaroon"`
+func (a *deviceAuthorizer) EnsureDeviceSession(dauthCtx DeviceAndAuthContext, client *http.Client) error {
+	if dauthCtx == nil {
+		return fmt.Errorf("internal error: no authContext")
 	}
 
-	headers := map[string]string{
-		"User-Agent":   snapdenv.UserAgent(),
-		"Accept":       "application/json",
-		"Content-Type": "application/json",
-	}
-	resp, err := retryPostRequestDecodeJSON(httpClient, MacaroonACLAPI, headers, macaroonJSONData, &responseData, nil)
+	device, err := dauthCtx.Device()
 	if err != nil {
-		return "", fmt.Errorf(errorPrefix+"%v", err)
+		return err
 	}
 
-	// check return code, error on anything !200
-	if resp.StatusCode != 200 {
-		return "", fmt.Errorf(errorPrefix+"store server returned status %d", resp.StatusCode)
+	if device.SessionMacaroon != "" {
```

Comment:
I would expect this to result in an error as well. Can you please add a short comment explaining why this is not treated as an error?

Reply:
it's checking that we have a SessionMacaroon  `(!= "")`, so there is nothing to do, it's not an error. I can add a comment about having nothing to do though.

Reply:
Ah, indeed, I misread it. But thanks for the comment!